src/main/scala/translating/GTIRBToIR.scala

```diff
@@ -0,0 +1,568 @@
+
+    val blockAddresses: mutable.Map[ByteString, Int] = mutable.Map.empty
+
+    for {
+      mod <- mods
+      section <- mod.sections
+      byteInterval <- section.byteIntervals
+      block <- byteInterval.blocks
+      if !block.getCode.uuid.isEmpty
+    } {
+      blockAddresses += (block.getCode.uuid -> (byteInterval.address + block.offset).toInt)
+    }
+
+    blockAddresses
+
+  }
+
+  // Creates CFG map mapping blk UUIDs to a buffer of outgoing edges 
+  def create_cfg_map(): (mutable.Map[String, ArrayBuffer[Edge]], mutable.Map[String, ArrayBuffer[Edge]]) = {
+
+    val edges = cfg.edges
+    val edgeMap: mutable.Map[String, ArrayBuffer[Edge]] = mutable.Map.empty
+    val incomingEdgeMap: mutable.Map[String, ArrayBuffer[Edge]] = mutable.Map.empty
+
+    for (edge <- edges) {
+
+      val targetUuid = Base64.getUrlEncoder.encodeToString(edge.targetUuid.toByteArray)
+      val sourceUuid = Base64.getUrlEncoder.encodeToString(edge.sourceUuid.toByteArray)
+
+      if (edgeMap.contains(sourceUuid)) {
+        edgeMap(sourceUuid) += edge
+      } else {
+        edgeMap += (sourceUuid -> ArrayBuffer(edge))
+      }
+
+      if (incomingEdgeMap.contains(targetUuid)) {
+        incomingEdgeMap(targetUuid) += edge
+      } else {
+        incomingEdgeMap += (targetUuid -> ArrayBuffer(edge))
+      }
+
+    }
+    (edgeMap, incomingEdgeMap)
+  }
+
+  // Another map of blk UUIDs to Symbols related to that blk. 
+  // Symbols are usually blk names or strings related to that blk
+  def create_symMap(): mutable.Map[ByteString, mutable.Set[Symbol]] = {
+    val symMap = mutable.Map[ByteString, mutable.Set[Symbol]]()
+
+    for (sym <- symbols) {
+      if (sym.optionalPayload.isReferentUuid) {
+        val ruuid = sym.optionalPayload.referentUuid.get
+
+        if (symMap.contains(ruuid)) {
+          symMap(ruuid) += sym
+        } else {
+          symMap += (ruuid -> mutable.Set[Symbol](sym))
+        }
+
+      }
+    }
+    symMap
+  }
+
+  // Gets name of function using a UUID
+  def create_names(uuid: ByteString): String = {
+    var name: String = uuid.toString()
+
+    if (functionNames.contains(uuid)) {
+      name = symbols.find(functionNames(uuid) == _.uuid).get.name
+
+      val entryBlocks: mutable.Set[ByteString] = functionEntries(uuid)
+
+      val result = entryBlocks
+        .flatMap(e => edgeMap.getOrElse(Base64.getUrlEncoder.encodeToString(e.toByteArray), None).iterator)
+        .flatMap(elem => mods.flatMap(_.proxies).find(_.uuid == elem.targetUuid))
+        .flatMap(proxy => symMap.getOrElse(proxy.uuid, None))
+        .map(p => p.name)
+
+      if (result.nonEmpty) {
+        return result.head //. head seems weird here but i guess it works
+      }
+
+    }
+
+    name
+  }
+
+  def get_jmp_reg(statement: Statement): Variable = statement match {
+    case LocalAssign(_, rhs: Variable, _) => rhs
+    case LocalAssign(_, rhs: Extract, _) => rhs.body.asInstanceOf[Variable]
+    case _ => ???
+  }
```

Comment:
This does not seem to be a reliable way to identify the register used in an indirect call. You should make sure it is an assignment to the program counter, rather than just assuming the most recent assignment is to the program counter.

Reply:
```scala 
def get_jmp_reg(statement: Statement): Variable = statement match {
    case LocalAssign(lhs: Register, rhs: Variable, _) if lhs.name == "_PC" => rhs
    case LocalAssign(lhs: Register, rhs: Extract, _) if lhs.name == "_PC" => rhs.body.asInstanceOf[Variable]
    case _ => ???
  }
```

Is something like this fine? If Gtirb doesn't put a write to _PC as the last statement to a block with an indirect call, than I think we're in the dark either way.

Reply:
Yeah, that would be better. If we encounter something strange like there not being a write to _PC before an indirect call, then it's best if it isn't handled so we can identify that there's a problem.